design/ceph/nvmeof-recoverer.md

```diff
@@ -0,0 +1,184 @@
+- **Minimized Downtime**: In traditional systems, node failures are manually recovered or reattached. The Rook operator retries pod-level recovery for transient failures, but if repeated restarts fail, it marks the OSD as out, allowing Ceph-level recovery (rebalancing). NVMe-oF recoverer minimizes this downtime by automatically reassigning storage devices to other nodes, ensuring data availability with minimal delay even in the event of unrecoverable compute node failures.
+
+- **Load balance**: As storage needs grow, Recoverer enables easy expansion of storage capacity. It introduces a scheduler that automatically connects fabric devices to suitable nodes using load-balancing and deploys OSDs. By utilizing Kubernetes Jobs, the scheduler dynamically connects based on the workload ensuring optimal performance
+
+### Non-Goals
+
+- **Support for Non-NVMe-oF Environments**: This design focuses solely on NVMe-oF environments.
+- **Modifications to the Core Ceph Codebase**: The proposal does not involve changing Ceph's core functionalities.
+
+## Proposal Details
+
+### Overview
+
+We propose a design, called NVMe-oF recoverer. NVMe-oF is an emerging storage network where the storage devices can be attached to any storage node using RDMA, TCP or Fiber channel. In this environment, the failure caused by a server node can be recoverable because the OSD can be re-spawned to another server node with the storage device attached to the failed server node through NVMe-oF.
+
+To address these challenges, we recommend a phased implementation consisting of four key steps:
+
+1. **Implement an Explicit Command for Manual OSD Relocation**
+2. **Automate OSD Relocation Upon Failure**
+3. **Modify CRUSH Map to Minimize Data Rebalancing**
+4. **Improve OSD Scheduling Based on Node Workloads**
+
+Each step can be implemented and contributed individually, allowing for incremental progress and testing.
+
+### Step 1: Implement an Explicit Command for Manual OSD Relocation
+
+#### Description
+
+Introduce explicit commands or scripts that allow administrators to manually deploy and relocate OSD pods using NVMe-oF devices. These tools will streamline the management of NVMe-oF devices and OSD pods in standard scenarios and during failures.
+
+#### Implementation Steps
+
+- **Develop a Script or Command-line Tool**:
+  - **For Connecting NVMe-oF Devices and Deploying OSD Pods**:
+    - Retrieve NVMe-oF target information from predefined sources (e.g., configuration files, environment variables).
+      - Key parameters: IP, Port, Subsystem NQN (SubNQN), and attachable nodes.
+    - Connect devices to target nodes using `nvme-cli` or appropriate tools.
+    - Update the `CephCluster` Custom Resource Definition (CRD) with fabric device details to facilitate automated OSD deployment.
+  - **For Handling OSD Relocation in Failure Scenarios**:
+    - Remove the faulted OSD pod.
```

Comment:
If the OSD is portable, it will not have specific node affinity, and we should rely on K8s scheduling to reschedule the pod on a down node. Rook shouldn't need any code for this scheduling.